Improve bash completion with compopt

sudotac opened the pull request:
This PR tries to improve bash completion support.

* Fix #5239
* Add support for `ValueHint::DirPath`
* Suppress a useless space appended during completion for `ValueHint::Other`

epage commented:
Could you insert a commit at the beginning that updates [`complete`](https://github.com/clap-rs/clap/blob/master/clap_complete/tests/testsuite/bash.rs#L149) for these cases and then in each commit as things get fixed, the asserts in that test can be updated?

We've added support for testing completions but it'd be a lot of work to go back and add them for all previous issues but I'd like for us to be making sure we cover them for future ones.

sudotac commented:
Moved the test change to https://github.com/clap-rs/clap/pull/5240/commits/6e1fae93288d31107a93df2543aa440839a80a7b. I hope I fixed this correctly.


> Also, will this cause problems for people on older bash? See https://github.com/clap-rs/clap/issues/5190.

I think it would work on bash 4.0 or later.
According to [the bash changelog (warning: very long)](https://git.savannah.gnu.org/cgit/bash.git/tree/CHANGES?h=bash-5.2), compopt was introduced in bash 4.0 and the options (`-o filenames`, `-o plusdirs` and `-o nospace`) are available on older versions.

IMO bash 4.0 is old enough to be a minimal supported version.
As far as I've searched on [repology](https://repology.org/project/bash/versions), no "major" distribution supports bash 4.0 or older. (It depends on what is "major" though...)

epage commented:
> In my opinion, bash 4.0 is old enough to be a minimum supported version.
As far as I've searched on [repology](https://repology.org/project/bash/versions), no "major" distribution supports bash 4.0 or older. (It depends on what is "major" though...)

The problem is macOS...

epage commented on clap_complete/tests/snapshots/home/static/exhaustive/bash/.bashrc:
```diff
@@ -542,6 +542,7 @@ _exhaustive() {
                     ;;
                 --other)
                     COMPREPLY=("${cur}")
+                    compopt -o nospace
```
> Moved the test change to https://github.com/clap-rs/clap/commit/6e1fae93288d31107a93df2543aa440839a80a7b. I hope I fixed this correctly.

I wasn't asking for the snapshots to be moved but functional tests to be added.

sudotac replied:
I completely misunderstood, sorry for bothering you.
I'll fix it.

sudotac replied:
Locally `cargo test` is fine, but CI failed because completion output is different.
Even if we remove 'nosort', the sort orders can be different (https://github.com/clap-rs/clap/pull/5026/commits/35a9d8438e2aa0d396a807b420df7752ee957293).

epage replied:
That is annoying.  nm then

sudotac replied:
> Locally cargo test is fine, but CI failed because completion output is different.

Fixed by checking the existence of file names and/or directory names instead of comparing snapshot of the output.

Also, created files and directories to be completed in a temporary directory with [`snapbox::path::PathFixture::mutable_temp`](https://docs.rs/snapbox/latest/snapbox/path/struct.PathFixture.html#method.mutable_temp) so that these tests won't be affected by future changes to this project.

sudotac commented:
> The problem is macOS...

Sorry, I confused it with macPorts. Then, this change will cause the same compatibility issue on macOS. It is necessary to check bash version.

epage commented on clap_complete/tests/testsuite/bash.rs:
```diff
@@ -160,6 +160,50 @@ fn complete() {
     snapbox::assert_eq(expected, actual);
+
+    // Issue 5239 (https://github.com/clap-rs/clap/issues/5239)
+    let input = "exhaustive hint --file test\t";
+    let expected = "exhaustive hint --file test     % exhaustive hint --file tests/";
```
If I understand correctly, this test fails in this commit and the later commits make it pass?

Can you update it so this commit passes, showing the current behavior, and the later commits update the tests to still pass as the behavior changes?
- We want `cargo test` to pass on every commit
- This helps show how each commit changes behavior

sudotac replied:
> If I understand correctly, this test fails in this commit and the later commits make it pass?

Yes.
Updated commits.

epage commented on clap_complete/src/shells/bash.rs:
```diff
@@ -164,29 +164,48 @@ fn option_details_for_path(cmd: &Command, path: &str) -> String {
+
+                if let Some(copt) = compopt {
+                    v.extend([
+                        r#"if [ "${BASH_VERSINFO[0]:-0}" -ge 4 ]; then"#.to_string(),
```
This compares the major version.  Do we also need to check the minor version like #5278 does?

sudotac replied:
I think we don't need to check the minor version because `compopt` was introduced in bash 4.**0**. (In other words, it is sufficient to check the major version only because the minor version is always greater than or equal to 0.)

By the way, should I implement this comparison in the same way of #5278 for consistency?
I wrote this in POSIX-compatible way considering #5011, but it requires much effort to fix everything...

epage replied:
I can go either way.  Without us having explicit testing for posix compatibility in our generated output, we'll regularly mess it up so I don't think its worth it until someone intentionally tries to support it.

sudotac replied:
OK. I'll go the same way as #5278.

sudotac commented:
I found out we cannot add tests for https://github.com/clap-rs/clap/pull/5240/commits/22bf919b71a621d380443cb0ef5e4b13a382b01f for now because white spaces are trimmed by runtime.complete().

epage commented:
> I found out we cannot add tests for [22bf919](https://github.com/clap-rs/clap/commit/22bf919b71a621d380443cb0ef5e4b13a382b01f) for now because white spaces are trimmed by runtime.complete().

Resolved in #5310

sudotac commented:
Thank you very much!
I've added a test for the remaining one.

epage commented on clap_complete/tests/testsuite/bash.rs:
```diff
@@ -161,6 +161,50 @@ fn complete() {
+            testdir_path.to_string_lossy()
+        );
+        let actual = runtime.complete(input.as_str(), &term).unwrap();
+        assert!(actual.contains("a_file"));
+        assert!(actual.contains("b_file"));
+        assert!(actual.contains("c_dir"));
+        assert!(actual.contains("d_dir"));
```
Is there a reason we aren't verifying the complete output in each of these cases?

I always worry about `contains` / `!contains` tests that things are showing up or not for the wrong reasons that is made clear by seeing the rest.

If there is a reason for us not to do so, we should at least show the full output in the assert message so people can debug the failure

sudotac replied:
Sorry for the late reply.

> Is there a reason we aren't verifying the complete output in each of these cases?

This is because the order of completion items is unstable (same as reported at https://github.com/clap-rs/clap/commit/35a9d8438e2aa0d396a807b420df7752ee957293).
Actually, the previous implementation passed in my local environment, but [failed on CI](https://github.com/clap-rs/clap/actions/runs/7196540335/job/19601627964#step:6:285).

> If there is a reason for us not to do so, we should at least show the full output in the assert message so people can debug the failure

Updated to show the full output of completion.
Additionally, collected multiple assertion calls into one.

epage commented:
Thanks!
